### PR TITLE
Refactor google playstore release flow

### PR DESCRIPTION
**Pull request by nid90**
## This addresses

- Extract PlayStore specific release functions to a separate class.
- Use the base `Installations::Error` for `Google::PlayDeveloper::Error` and add reasons for all matched errors.
- Deprecate `upload_failed` terminal state for DeploymentRun, use `failed` with a proper `failure_reason` instead.
- Add failure reason to `release_failed` passport stamp.
- Add fully release action for PlayStore. Move the action from DeploymentRun to StagedRollout.
- Add a new state in StagedRollout to handle the full release in the middle of a staged rollout.
- Minor bugfix: do not show halt action on UI till the staged rollout has moved to at least the first stage.

UI change to allow fully release: 

Screenshot 2023-03-14 at 7 15 54 PM

Deployment (and staged rollout) in the UI after fully released:

Screenshot 2023-03-14 at 7 18 50 PM

### CODE REVIEW

File: app/libs/deployments/google_play_store/release.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        end
+      end
+
+      def release_to_all!(&blk)
+        return unless google_play_store_integration?
+
+        release_with(rollout_value: Deployment::FULL_ROLLOUT_VALUE, &blk)
+      end
+
+      # TODO: handle known errors gracefully and show to users
```
Comment (kitallis) [inline app/libs/deployments/google_play_store/release.rb:95]: Is this `TODO` still necessary?

**Review by kitallis — COMMENTED** [review]
I think the indirections are getting a bit out of hand, let's discuss.

File: app/libs/deployments/google_play_store/release.rb (hunk 1 above)
Comment (nid90) [reply app/libs/deployments/google_play_store/release.rb:95]: the graceful handling is happening, but we are still not showing it to users.

I wanted to add a small UI change to show the human-readable `failure_reason` for when a deployment fails whether via manual trigger or a job.

File: db/data/20230314084413_deprecate_upload_failed_status.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class DeprecateUploadFailedStatus < ActiveRecord::Migration[7.0]
+  def up
```
[the remark this answers is not in the document]
Comment (nid90) [reply db/data/20230314084413_deprecate_upload_failed_status.rb:2]: oh. that seems too defensive, no?

**Comment by nid90** [comment]
> I think the indirections are getting a bit out of hand, let's discuss.

This is what, in part, was bothering me about `halt` flow. It is becoming a little more obvious now with `fully_release` flow.

I have cleaned up the yields though and named the functions which yield a block better.

File: app/libs/deployments/google_play_store/release.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        else
+          fully_release!
+        end
+      end
+
+      def halt_release!
+        return unless google_play_store_integration?
+
+        return unless run.rollout_started?
+        provider.halt_release(deployment_channel, build_number, release_version, run.staged_rollout.last_rollout_percentage)
```
Comment (kitallis) [inline app/libs/deployments/google_play_store/release.rb:82]:
```suggestion
        return unless google_play_store_integration?
        return unless run.rollout_started?
        
        provider.halt_release(deployment_channel, build_number, release_version, run.staged_rollout.last_rollout_percentage)
```

File: app/libs/deployments/google_play_store/release.rb [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+        run.fail_with_error(result.error) unless result.ok?
+        result
+      end
+
+      private
+
+      def fully_release!
+        result = provider.rollout_release(deployment_channel, build_number, release_version, Deployment::FULL_ROLLOUT_VALUE)
+        if result.ok?
```
Comment (kitallis) [inline app/libs/deployments/google_play_store/release.rb:107]:
```suggestion
        result = provider.rollout_release(deployment_channel, build_number, release_version, Deployment::FULL_ROLLOUT_VALUE)
        
        if result.ok?
```

File: app/libs/deployments/google_play_store/release.rb [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+          run.complete!
+        else
+          run.fail_with_error(result.error)
+        end
+      end
+
+      def rollout!
+        result = provider.create_draft_release(deployment_channel, build_number, release_version)
+        if result.ok?
+          run.create_staged_rollout!(config: staged_rollout_config)
```
Comment (kitallis) [inline app/libs/deployments/google_play_store/release.rb:117]:
```suggestion
        result = provider.create_draft_release(deployment_channel, build_number, release_version)
        
        if result.ok?
          run.create_staged_rollout!(config: staged_rollout_config)
```

**Review by kitallis — APPROVED** [review]

**Pull request merged**